table/block.cc

```diff
@@ -285,10 +302,20 @@ void BlockIter::CorruptionError() {
   value_.clear();
 }
 
-bool DataBlockIter::ParseNextDataKey() {
+// if within_restart_interval == true, we only parse next key within
+// current restart invterval. The default value of within_restart_interval
+// is false.
+bool DataBlockIter::ParseNextDataKey(bool within_restart_interval) {
   current_ = NextEntryOffset();
   const char* p = data_ + current_;
   const char* limit = data_ + restarts_;  // Restarts come right after data
+  if (within_restart_interval) {
+    assert(restart_index_ < num_restarts_);
+    if (restart_index_ + 1 < num_restarts_) {
+      limit = data_ + GetRestartPoint(restart_index_ + 1);
```

Comment:
I don't understand why the logic happens here, not in HashSeek(). Why do you need to recalculate limit when iterating every entry?